### PR TITLE
Preview for #174 and #187 - Selecting and filtering sources and resources

### PR DESCRIPTION
**Pull request by abelsromero**
**PR description**
This PR is a preview of a solution for issues #174 and #187.
Both issues highlighted the limitations in terms of sources and resources management of the maven plugin.
Following the recomendation of Dan on isue #174
```
Agreed. We should align with the Gradle plugin on this one.
See https://github.com/asciidoctor/asciidoctor-gradle-plugin#usage (sources)
```
this PR includes the following changes:
(1). `sourceDirectory`, `sourceDocumentName` & `sourceDocumentExtension` disappear in favour of two new options sources & resources.
(2). `sources`: follows a the same resource pattern as the [maven-resource-plugin](https://maven.apache.org/plugins/maven-resources-plugin/examples/include-exclude.html). Basically it uses the same underlying API (https://maven.apache.org/shared/maven-filtering/). The difference here is that filtering/replacement options are not available and by default only AsciiDoc files are included.
Rendered files can be modified using custom include and exclude filters. This includes, adding more sources files with custom file extensions, adding multiple files, directories, etc.
(3). `resources`: follows the same resources pattern [...]. If not set, all resources (images, css, etc.) in the different sources are copied. This is different to how the gradle plugin works, but is consistent with how the current maven plugn works. Also, all source files defined using `include` conditions are excluded by default. In this case, the filtering (a.k.a. replacement) options provided by the maven-resource-plugin can be used, I would not advertise them.
All of this can be modified using custom include and exclude filters. For instance, if someone wants to copy AsciiDoc documents to the target folder, just create a <resource> tag like this one:
```
<resource>
    <directory>DIRECTORY</directory>
    <include>**/*.adoc</>
    <include>**/*.asciidoc</>
</resource>
```

(4). Options related to the output (format and directory) are kept 'global', that is, at <configuration> level as now. To some degree it makes sense to define them in the source element, but in my opinion this is just re-implementing maven executions inside the plugin. This includes options like `outputDirectory`, `baseDir`, `backend`, `preserveDirectories`, etc.
The criteria I followed was:
 * You need to manage several inputs? Now you can :)
 * You need to manage multiple outputs? You should use different executions. In the same way in gradle you would define a different task.
But eventually, this could be implemented if the community asks for it. But this will mean that resources will a property inside each source.

**Next steps**
* Le'ts discuss this and please let me know of corner cases to test. There's quite a lot of include/exclude manipulation and I suspect I may have slipt some special case.
I have separated working code and test to ease the review. I'll merge them on the final PR.
* Documentation is missing, once we have the final approach I'll include it.
* Decide version: this is a major change with no backward compatibility. Should we save it for 1.6.0 and use 1.5.4/5 for urgent fixes?
* Improve and document the refresh mojo: right now I did a quick approach 'that works' but can consume lots of resources in some cases. I keep that for a separated PR.

* @marcaurele, to render the files in the root (README, CONTRIBUTING, etc.), you can do something like this:
```
<sources>
    <source>
        <directory>.</directory>
        <include>*.adoc</include> <!-- only includes files in 'directory' property, not sub-directories -->
    </source>
</sources>
<resources>
    <resource>
        <directory>.</directory>
        <exclude>**/**</exclude> <!-- does not copy anything -->
    </resource>
</resources>
```
or include each file one by one with differnt include tags. _Let me lnow what do you think_

* @AlexanderZobkov please, let me know if this approach helps you.

_I think that's all, thanks for you patience if you got here_

### CODE REVIEW

**Comment by abelsromero** [comment]
Build fails in the CIs because integration test are not updated. I'll add that to the list of pending items.

**Comment by marcaurele** [comment]
I got an error with this config:

```
					<sources>
						<directory>.</directory>
						<include>*.adoc</include>
					</sources>
					<resources>
						<directory>.</directory>
						<exclude>**/**</exclude>
					</resources>

```

```
doctor-maven-plugin:1.5.4-SNAPSHOT:process-asciidoc (generate-html-doc) on project sample: Unable to parse configuration of mojo org.asciidoctor:asciidoctor-maven-plugin:1.5.4-SNAPSHOT:process-asciidoc for parameter directory: Cannot find default setter in class org.apache.maven.model.Resource
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:220)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:317)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:152)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:555)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:214)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:158)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginConfigurationException: Unable to parse configuration of mojo org.asciidoctor:asciidoctor-maven-plugin:1.5.4-SNAPSHOT:process-asciidoc for parameter directory: Cannot find default setter in class org.apache.maven.model.Resource
        at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:602)
        at org.apache.maven.plugin.internal.DefaultMavenPluginManager.getConfiguredMojo(DefaultMavenPluginManager.java:534)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:97)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 19 more
Caused by: org.codehaus.plexus.component.configurator.ComponentConfigurationException: Cannot find default setter in class org.apache.maven.model.Resource
        at org.eclipse.sisu.plexus.CompositeBeanHelper.setDefault(CompositeBeanHelperjava:95)
        at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.fromConfiguration(ObjectWithFieldsConverter.java:61)
        at org.codehaus.plexus.component.configurator.converters.composite.AbstractCollectionConverter.fromChildren(AbstractCollectionConverter.java:54)
        at org.codehaus.plexus.component.configurator.converters.composite.CollectionConverter.fromConfiguration(CollectionConverter.java:68)
        at org.eclipse.sisu.plexus.CompositeBeanHelper.convertPropertyCompositeBeanHelper.java:268)

        at org.eclipse.sisu.plexus.CompositeBeanHelper.setProperty(CompositeBeanHelper.java:174)
        at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.processConfiguration(ObjectWithFieldsConverter.java:101)
        at org.codehaus.plexus.component.configurator.BasicComponentConfigurator.configureComponent(BasicComponentConfigurator.java:32)
        at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:572)
        ... 22 more
```

**Comment by abelsromero** [comment]
Right now it's no published, to you have to install it manually:
* Clone my repo: https://github.com/abelsromero/asciidoctor-maven-plugin/
* Switch to branch: issues_174_187
https://github.com/abelsromero/asciidoctor-maven-plugin/
* Run a `mvn install` to generate the jar and publish it into you local repo.

Tomorrow I can try to publish it on my Bintray tray to make things easier.

**Comment by abelsromero** [comment]
I got it done much quicker than I thought.
* My repo: http://dl.bintray.com/abelsromero/maven/
* Version: 1.5.4-preview.1

Full info: https://bintray.com/abelsromero/maven/asciidoctor-maven-plugin/1.5.4-preview.1/

If you have problems setting the project, the POM of the project has an example of how to set an additional repo with a profile: https://github.com/asciidoctor/asciidoctor-maven-plugin/blob/master/pom.xml#L374-L388

**Comment by marcaurele** [comment]
I did checkout your branch and run <code>mvn install</code> before. Then I updated the version of the ascii-doctor plugin in my pom to 1.5.4-SNAPSHOT and change the configuration with the 2 new sections. Maybe the configuration is not written correctly. From your comment, I understood I should write the configuration like that:
```xml
		<configuration>
			<!--sourceDocumentName>../../../README.adoc</sourceDocumentName-->
			<preserveDirectories>false</preserveDirectories>
                        <headerFooter>true</headerFooter>
                        <!-- Attributes common to all output formats -->
                        <attributes>
                            <docVersion>${project.version}</docVersion>
                            <icons>font</icons>
                            <toc />
                            <sectanchors />
                            <idprefix />
                            <idseparator>-</idseparator>
                        </attributes>
			<sources>
				<directory>.</directory>
				<include>*.adoc</include>
			</sources>
			<resources>
				<directory>.</directory>
				<exclude>**/**</exclude>
			</resources>
		</configuration>
```

Is it correct?

**Comment by abelsromero** [comment]
My explanation was crap, sorry.
I have created a demo project with all details. Just clone it and run `mvn`: https://github.com/abelsromero/asciidoctor-maven-plugin-preview/

The example already points to the precompiled version on bintray, contains two executions showing different ways to render documents in the root path and many documents and folders to play with.

Note that you can have multiple <source> and <resource> blocks if you need to.

**Comment by abelsromero** [comment]
Any comment on this? @mojavelinux, @LightGuard ?

**Comment by LightGuard** [comment]
TL;DR +1 for the idea and approach, and also putting it in 1.6.0. -1 for
1.5.4 unless we still support a deprecated current approach alongside this
change.

I've not played with this, but I like the change and the approach seems
sound. It does feel like at 1.6.0 change though as we're changing how you
use the plugin and we're not keeping a backwards compatible usage. If we
could do this, deprecate the current approach (but still honor it) and
introduce this all in one go I'd be happy with that. We need to let people
know about the change and give them an upgrade path and time to make that
upgrade.

-- 
Sent from Gmail Mobile

**Comment by abelsromero** [comment]
> support a deprecated current approach alongside this change.

We could support `sourceDirectory` & `sourceDocumentExtension` as configuration shortcuts. But then it would be possible to set the same folder as a <source> and is sourceDirectory, which raises some questions of what should be the behauviour. But still, it could be doable.

> keeping a backwards compatible
I am not in favour of keeping backward compatibility in these case. The changes are big and I think I'll end up keeping an execution workflow for the old code and a totally separated on for the new one. Something like a big `if-else`.
Dan wanted to do a 1.5.5 release, I think we can start announcing these changes, publish a RC and schedule them for the 1.6.0 release.

**Comment by AlexanderZobkov** [comment]
@abelsromero Finally, I was able to look on it.
The approach with `sources` works for me. Thank you!
Regarding `resources` and filtering capabilities, what delimiters for expresssions to be filtered will be recognized, ${xxx} or @{yyy} or something else?

IMHO, this is definitely a major change and should be introduced:
- in 1.6.0 release with no backward compatability with old configuration
or 
- in 1.5.5 release but with backward compatability with old configuration

Regarding the plugin documentation, it would be great to see since when this or that configuration option was introduced. BTW, why standard maven plugin documentation generator is not used?

**Comment by abelsromero** [comment]
@AlexanderZobkov, glad it serves you.

> IMHO, this is definitely a major change and should be introduced:
> 
> in 1.6.0 release with no backward compatability with old configuration or
> in 1.5.5 release but with backward compatability with old configuration

I think we can agree this going to be for 1.6.0. To ease the change, sourceDirectory & sourceDocumentExtension can be supported without trouble. My impression is that most builds are based on rendering a single folder, so those using these options won't need to change anything in their pom's.
But I don't feel like creating a whole intermediate 1.5.5 version with full compatibilty.

> Regarding resources and filtering capabilities, what delimiters for expresssions to be filtered will be recognized, ${xxx} or @{yyy} or something else?

If you are you thinking in things like ${baseDir}, so far none, but havent' looked much into it. I saved that for a future relase once the approach is agreed.

> Regarding the plugin documentation, it would be great to see since when this or that configuration option was introduced. BTW, why standard maven plugin documentation generator is not used?

We have been rolling out eveything on GitHub, focusing on the README and referencing current version options, but nothing prevents us from doing this in the future. For instance, AsciidoctorJ has been working to publish the javadocs which weren't available untill now...You know, step by step.

**Comment by abelsromero** [comment]
After giving some though to these changes, I think I'll split them in several minnor PR to ease the integration, review and compatibility.
I'll start adding the option to set resources to copy and after that...I'll see :P

**Comment by mojavelinux** [comment]
I'm really excited about this change. I think it brings both simplicity and power to the plugin, which is an amazing feat. It also helps with support that it mostly aligns with the Gradle plugin to the degree it makes sense. Really nice work! I look forward to the follow-ups. Perhaps target a new 1.6.0 branch?

**Comment by abelsromero** [comment]
My immediate plan is to add the resource option since it makes sense on its own and then release a 1.5.4 version of the plugin. After that, well I really haven't thought much :P
